circuit-breaker/src/main/java/com/iluwatar/circuitbreaker/DefaultCircuitBreaker.java

```diff
@@ -0,0 +1,153 @@
+   *                         state to 'OPEN'
+   * @param retryTimePeriod  Time period after which a new request is made to remote service for
+   *                         status check.
+   */
+  DefaultCircuitBreaker(RemoteService serviceToCall, long timeout, int failureThreshold,
+      long retryTimePeriod) {
+    this.service = serviceToCall;
+    // We start in a closed state hoping that everything is fine
+    this.state = State.CLOSED;
+    this.failureThreshold = failureThreshold;
+    // Timeout for the API request.
+    // Used to break the calls made to remote resource if it exceeds the limit
+    this.timeout = timeout;
+    this.retryTimePeriod = retryTimePeriod;
+    //An absurd amount of time in future which basically indicates the last failure never happened
+    this.lastFailureTime = System.nanoTime() + futureTime;
+    this.failureCount = 0;
+  }
+
+  //Reset everything to defaults
+  @Override
+  public void recordSuccess() {
+    this.failureCount = 0;
+    this.lastFailureTime = System.nanoTime() + futureTime;
+    this.state = State.CLOSED;
+  }
+
+  @Override
+  public void recordFailure() {
+    failureCount = failureCount + 1;
+    this.lastFailureTime = System.nanoTime();
+  }
+
+  //Evaluate the current state based on failureThreshold, failureCount and lastFailureTime.
+  protected void evaluateState() {
+    if (failureCount >= failureThreshold) { //Then something is wrong with remote service
+      if ((System.nanoTime() - lastFailureTime) > retryTimePeriod) {
+        //We have waited long enough and should try checking if service is up
+        state = State.HALF_OPEN;
+      } else {
+        //Service would still probably be down
+        state = State.OPEN;
+      }
+    } else {
+      //Everything is working fine
+      state = State.CLOSED;
+    }
+  }
+
+  @Override
+  public String getState() {
+    evaluateState();
+    return state.name();
+  }
+
+  /**
+   * Break the circuit beforehand if it is known service is down Or connect the circuit manually if
+   * service comes online before expected.
+   *
+   * @param state State at which circuit is in
+   */
+  @Override
+  public void setState(State state) {
+    this.state = state;
+    switch (state) {
+      case OPEN:
+        this.failureCount = failureThreshold;
+        this.lastFailureTime = System.nanoTime();
+        break;
+      case HALF_OPEN:
+        this.failureCount = failureThreshold;
+        this.lastFailureTime = System.nanoTime() - retryTimePeriod;
+        break;
+      default:
+        this.failureCount = 0;
+    }
+  }
+
+  /**
+   * Executes service call.
+   *
+   * @return Value from the remote resource, stale response or a custom exception
+   */
+  @Override
+  public String attemptRequest() throws RemoteServiceException {
+    evaluateState();
+    if (state == State.OPEN) {
+      // return  cached response if no the circuit is in OPEN state
+      return "This is stale response from API";
```

Comment:
Instead of hard coded string, can we actually cache the response and return it?

Reply:
Yes that would be better. Fixed this in new commit